File: problem2.cpp
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
public:
    int d1,d2;
    TreeNode* p1,*p2;

    void solve(TreeNode* root,int depth,TreeNode* parent,int x,int y){
        if(!root)
        return ;

        depth+=1;
        parent=root;

        if(root->left){
            if(root->left->val==x){
            d1=depth;
            p1=root;
            return;
            } else if(root->left->val==y){
                d2=depth;
                p2=root;
            } else
                solve(root->left,depth,parent,x,y);
            

        }

        if(root->right){
            if(root->right->val==y){
            d2=depth;
            p2=root;
            return;
            } else if(root->right->val==x){
            d1=depth;
            p1=root;
            return;
            } else
                solve(root->right,depth,parent,x,y);
            

        }        
    }

    bool isCousins(TreeNode* root, int x, int y) {
        solve(root,0,NULL,x,y);
        if(d1==d2 and p1!=p2)
        return 1;

        return 0;
        
    }
};

File: Problem2.cs
public class CousinsInBT
    {
        // Time Complexity : O(n)
        // Space Complexity : O(h)
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        TreeNode parent_x;
        TreeNode parent_y;
        int depth_x;
        int depth_y;

        public bool IsCousins(TreeNode root, int x, int y)
        {
            dfs(root, null, x, y, 0);
            return depth_x == depth_y && parent_x != parent_y;

        }

        public void dfs(TreeNode root, TreeNode parent, int x, int y, int depth)
        {
            //base case
            if (root == null) return;

            //logic
            if(root.val == x)
            {
                parent_x = parent;
                depth_x = depth;
            }
            if(root.val == y)
            {
                parent_y = parent;
                depth_y = depth;
            }

            if(parent_x == null || parent_y == null)
            {
                dfs(root.left, root, x, y, depth + 1);
                dfs(root.right, root, x, y, depth + 1);
            }


        }
    }
